driver-core/src/main/com/mongodb/internal/connection/tlschannel/util/Util.java

```diff
@@ -44,15 +48,24 @@ public static String resultToString(SSLEngineResult result) {
   }
 
   public static int getJavaMajorVersion() {
-    String version = System.getProperty("java.version");
-    if (version.startsWith("1.")) {
-      version = version.substring(2, 3);
-    } else {
-      int dot = version.indexOf(".");
-      if (dot != -1) {
-        version = version.substring(0, dot);
+      return getJavaMajorVersion(System.getProperty("java.version"));
+  }
+
+  @VisibleForTesting(otherwise = PRIVATE)
+  public static int getJavaMajorVersion(final String javaVersion) {
+      String version = javaVersion;
+      if (version.startsWith("1.")) {
+          version = version.substring(2);
       }
-    }
-    return Integer.parseInt(version);
+      // Allow these formats:
+      // 1.8.0_72-ea
+      // 9-ea
+      // 9
+      // 9.0.1
+      // 17
+      int dotPos = version.indexOf('.');
+      int dashPos = version.indexOf('-');
+      return Integer.parseInt(
+              version.substring(0, dotPos > -1 ? dotPos : dashPos > -1 ? dashPos : version.length()));
```

Comment:
The only way we deviate here from the algorithm in tls-channel is that instead of using `1` at the end of this [line](https://github.com/marianobarrios/tls-channel/blob/42acba44f23cf5c8969ced85c8331346ee8d85da/src/main/java/tlschannel/util/Util.java#L44), we use `version.length()`. Without this change `"17"` is parsed as 1, which then causes class loading issues with the root cause being swallowed and not being present in the logs.

I'll create a PR for tls-channel when this PR is approved.